### ISSUE TITLE
pre-decrement-expression in exception filter wrong compilation

### ISSUE THREAD

**Issue by alekole** [dotnet/roslyn#7030]
repeated while compile C# in VS 2015 on windows 7 when flag "optimize code" checked.

 Create in empty asp.net website application (i try .net 4.5.2 and .net 4.6). add this code to global application class

    public class Global : System.Web.HttpApplication
    {
        protected void Application_Start(object sender, EventArgs e)
        {
            int i = 3;
            do
            {
                try
                {
                    throw new Exception();
                }
                catch (Exception) when (--i < 0)
                {
                    throw;
                }
                catch (Exception)
                {
                    Thread.Sleep(10);
                }
            } while (true);
        }
    }

Project properties -> Build -> check "optimize code". Run application. Get Infinite loop.

IL of  Application_Start method:

    .method family hidebysig 
	instance void Application_Start (
		object sender,
		class [mscorlib]System.EventArgs e
	) cil managed 
    {
	// Method begins at RVA 0x2050
	// Code size 48 (0x30)
	.maxstack 2
	.locals init (
		[0] int32 i
	)

	IL_0000: ldc.i4.3
	IL_0001: stloc.0
	// loop start (head: IL_0002)
		IL_0002: nop
		.try
		{
			IL_0003: newobj instance void [mscorlib]System.Exception::.ctor()
			IL_0008: throw
		} // end .try
		filter
		{
			IL_0009: isinst [mscorlib]System.Exception
			IL_000e: dup
			IL_000f: brtrue.s IL_0015

			IL_0011: pop
			IL_0012: ldc.i4.0
			IL_0013: br.s IL_0021

			IL_0015: pop
			IL_0016: ldloc.0
			IL_0017: ldc.i4.1
			IL_0018: sub
			IL_0019: dup
			IL_001a: stloc.0
			IL_001b: ldc.i4.0
			IL_001c: clt
			IL_001e: ldc.i4.0
			IL_001f: cgt.un

			IL_0021: endfilter
		} // end filter
		catch
		{
			IL_0023: pop
			IL_0024: rethrow
		} // end handler
		catch [mscorlib]System.Exception
		{
			IL_0026: pop
			IL_0027: ldc.i4.s 10
			IL_0029: call void [mscorlib]System.Threading.Thread::Sleep(int32)
			IL_002e: leave.s IL_0002
		} // end handler
	// end loop
    } // end of method Global::Application_Start

As i undestand it's not copying calculated value back to i variable

Without "optimize code" works fine, in console app - works fine.

Microsoft Visual Studio Professional 2015
Version 14.0.23107.0 D14REL
Microsoft .NET Framework
Version 4.6.00081

Installed Version: Professional
Visual C# 2015   00325-60000-46632-AA934
Microsoft Visual C# 2015
ASP.NET and Web Tools   14.0.20626.0
ASP.NET and Web Tools

**Comment by leppie** [comment]
The following:
```
using System;

public class Program
{
    static void Main(string[] args)
    {
        int i = 3;
        do
        {
            try
            {
                throw new Exception();
            }
            catch (Exception) when (--i < 0)
            {
                throw;
            }
            catch (Exception)
            {
                Console.WriteLine(i);
            }
        } while (true);
    }
}
```
Prints:
```
2
1
0

Unhandled Exception: System.Exception: Exception of type 'System.Exception' was thrown.
   at Program.Main(String[] args)
```
as expected. Not sure what behavior you are expecting other than this.

**Comment by alekole** [comment]
well  I  sad about web application, in console it works well

**Comment by leppie** [comment]
How do you know it is an infinite loop? The throw will kill the webapp and it probably just starts again and then kills itself again, etc , etc.

**Comment by alekole** [comment]
when (--i < 0)  this expression is never "true", so exception catched in next block, and then start another loop.

**Comment by leppie** [comment]
I cant repro on web, I get the error page as expected. You likely not seeing changes to `i` due to having optimize code turned on. Put a `Debugger.Break()` in the `catch` with a filter.

**Comment by gafter** [comment]
@alekole Does the following program *also* appear to result in an infinite loop for you?

```cs
    public class Global : System.Web.HttpApplication
    {
        protected void Application_Start(object sender, EventArgs e)
        {
            Thread.Sleep(10);
            Thread.Sleep(10);
            Thread.Sleep(10);
            Thread.Sleep(10);
            throw new Exception();
        }
    }
```

**Comment by alekole** [comment]
Non, this code throw "System.Exeption"

**Comment by jaredpar** [comment]
Closing for now as we're unable to reproduce this bug.  Going to submit a PR for a regression test.

**Comment by jaredpar** [comment]
Glad I decided to write the regression test, doing so caused me to stumble into a repro here.  This is the simplified code I wrote for the unit test:

``` csharp
using System;

class C
{
    static void Main()
    {
        int i = 3;
        do
        {
            try
            {
                throw new Exception();
            }
            catch (Exception) when (--i < 0)
            {
                Console.Write(""e"");
                break;
            }
            catch (Exception)
            {
                Console.Write(""h"");
            }
        } while (true);
    }
}
```

Testing showed that this will fail when the code is:

- Compiled with optimizations enabled (release mode)
- Run in a 64 bit environment.

Every other combo and the code will succeed, 

Moving to @VSadov

**Comment by VSadov** [comment]
The IL looks correct. 

```IL
	IL_0001: stloc.0
	// loop start (head: IL_0002)
		IL_0002: nop
		.try
		{
			IL_0003: newobj instance void [mscorlib]System.Exception::.ctor()
			IL_0008: throw
		} // end .try
		filter
		{
			IL_0009: isinst [mscorlib]System.Exception
			IL_000e: dup
			IL_000f: brtrue.s IL_0015

			IL_0011: pop
			IL_0012: ldc.i4.0
			IL_0013: br.s IL_0021

			IL_0015: pop
			IL_0016: ldloc.0                    <-- load i
			IL_0017: ldc.i4.1
			IL_0018: sub                        <-- decrement 
			IL_0019: dup
			IL_001a: stloc.0                    <-- store i , are we missing this store in x64?
			IL_001b: ldc.i4.0
			IL_001c: clt
			IL_001e: ldc.i4.0
			IL_001f: cgt.un

			IL_0021: endfilter
		} // end filter
		catch
		{
			IL_0023: pop
			IL_0024: ldstr "e"
			IL_0029: call void [mscorlib]System.Console::Write(string)
			IL_002e: leave.s IL_003d
		} // end handler
		catch [mscorlib]System.Exception
		{
			IL_0030: pop
			IL_0031: ldstr "h"
			IL_0036: call void [mscorlib]System.Console::Write(string)
			IL_003b: leave.s IL_0002
		} // end handler
	// end loop
```

**Comment by VSadov** [comment]
It seems to be a JIT bug specific to 64bit.
From looking at x64 it seems that we are missing a lot of code related to i. It is basically completely elided.

My theory is that the JIT misses the store in the filter condition (--i < 0). Since that is the only store, it concludes that i is a constant.

**Comment by gafter** [comment]
We now have a regression test for this in #10216.

**Comment by gafter** [comment]
The CLR bug is https://github.com/dotnet/coreclr/issues/4044

**Comment by JosephTremoulet** [comment]
The CLR bug has now been fixed (in CoreCLR) with dotnet/coreclr#4591 / dotnet/coreclr@0029475f1b
